### PR TITLE
Modified EHJob controller for dynamic scaling and to utilize PodGroup + Coscheduling Plugin

### PR DESCRIPTION
**Pull request by muffin-rice**
Added: 

- Code to check if TargetReplicas is changed while the job is running and change the StatefulSet accordingly 
- Sleep command option for convenient debugging 
- Replace directive to go.mod for compatability with sigs.k8s.io/scheduler-plugins import 
- Creation of PodGroup to ElasticHorovodJob controller 
- PodGroup name label to PodSpec of StatefulSet workers
- Code to check if coscheduling plugin rejects pods and take action accordingly (decrease number of Replicas by 1)

### CODE REVIEW

File: elastic-training/controllers/elastichorovodjob_controller.go [hunk 1]
```diff
@@ -195,16 +296,20 @@ func (r *ElasticHorovodJobReconciler) desiredWorkers(ehjob aiv1alpha1.ElasticHor
...
 			Template: corev1.PodTemplateSpec{
-				ObjectMeta: metav1.ObjectMeta{
-					Labels: map[string]string{"ElasticHorovodJob": ehjob.Name, "role": "worker"},
+				ObjectMeta: metav1.ObjectMeta{ //add labels such as podgroup # and statefulset
+					Labels: map[string]string{"ElasticHorovodJob": ehjob.Name, "role": "worker",
+						"pod-group.scheduling.sigs.k8s.io": fmt.Sprintf("podgroup-ehjob-%s", ehjob.Name),
+					},
 				},
 				Spec: corev1.PodSpec{
+					SchedulerName: "default-scheduler",
```
Comment (hxhp) [inline elastic-training/controllers/elastichorovodjob_controller.go:312]: Is this necessary? I assume we are using the default scheduler anyway.

File: elastic-training/controllers/elastichorovodjob_controller.go [hunk 2]
```diff
@@ -215,6 +320,15 @@ func (r *ElasticHorovodJobReconciler) desiredWorkers(ehjob aiv1alpha1.ElasticHor
 								},
 							},
 						},
+						//shared memory
```
Comment (hxhp) [inline elastic-training/controllers/elastichorovodjob_controller.go:323]: What is the use for this shared memory?

File: elastic-training/controllers/elastichorovodjob_controller.go [hunk 3]
```diff
@@ -288,12 +402,14 @@ func (r *ElasticHorovodJobReconciler) desiredLauncherJob(ehjob aiv1alpha1.Elasti
 	scriptMode := int32(0744)
 	one := int32(1)
 	sshSetupCommand := "mkdir -p /root/.ssh; cp /etc/secrets/* /root/.ssh/; chmod 644 /root/.ssh/authorized_keys;"
-	scriptSetupCmd := fmt.Sprintf("mkdir -p /scripts; cp /etc/scripts/* /scripts/; sed -i 's/SERVICENAME/%s/' /scripts/discover_hosts.sh;", svcName)
+	scriptSetupCmd := fmt.Sprintf("mkdir -p /scripts; mkdir -p /elastic_scripts; cp /etc/scripts/* /scripts/; sed -i 's/SERVICENAME/%s/' /scripts/discover_hosts.sh;", svcName)
```
Comment (hxhp) [inline elastic-training/controllers/elastichorovodjob_controller.go:405]: I don't see this /elastic_scripts used anywhere?

File: elastic-training/controllers/elastichorovodjob_controller.go [hunk 4]
```diff
@@ -288,12 +402,14 @@ func (r *ElasticHorovodJobReconciler) desiredLauncherJob(ehjob aiv1alpha1.Elasti
...
 	horovodArgs := fmt.Sprintf("-np %d --max-np %d --host-discovery-script /scripts/discover_hosts.sh",
 		*ehjob.Spec.WorkersSpec.MinReplicas,
 		*ehjob.Spec.WorkersSpec.MaxReplicas)
 	pythonCommand := strings.Join(ehjob.Spec.LauncherSpec.PythonCommand, " ")
-	wholeCommand := fmt.Sprintf("%s %s horovodrun -p 12345 %s %s", sshSetupCommand, scriptSetupCmd, horovodArgs, pythonCommand)
+	//sleepCommand := "sleep 600;"
+	sleepCommand := ""
```
Comment (hxhp) [inline elastic-training/controllers/elastichorovodjob_controller.go:411]: This is launcher pod, not sure why a sleep command is needed here.

File: elastic-training/controllers/elastichorovodjob_controller.go [hunk 5]
```diff
@@ -34,6 +36,8 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	aiv1alpha1 "elastictraining/api/v1alpha1"
+
+	sigsv1alpha1 "sigs.k8s.io/scheduler-plugins/pkg/apis/scheduling/v1alpha1"
```
Comment (hxhp) [inline elastic-training/controllers/elastichorovodjob_controller.go:40]: So it's using the official repo, not your forked repo?

Comment (muffin-rice) [reply elastic-training/controllers/elastichorovodjob_controller.go:40]: Yes, I am using https://pkg.go.dev/sigs.k8s.io/scheduler-plugins now.

File: elastic-training/controllers/elastichorovodjob_controller.go (hunk 2 above)
Comment (muffin-rice) [reply elastic-training/controllers/elastichorovodjob_controller.go:322]: PyTorch Horovod seems to require a shared memory mount

File: elastic-training/controllers/elastichorovodjob_controller.go (hunk 1 above)
Comment (muffin-rice) [reply elastic-training/controllers/elastichorovodjob_controller.go]: I added default-scheduler for easier changing if the coscheduling scheduler is configured to another name; comment added

File: elastic-training/controllers/elastichorovodjob_controller.go (hunk 3 above)
Comment (muffin-rice) [reply elastic-training/controllers/elastichorovodjob_controller.go]: Removed

**Review by hxhp — APPROVED** [review]
LGTM

**Pull request merged**